planning/scenario_selector/src/scenario_selector_node/scenario_selector_node.cpp

```diff
@@ -88,19 +88,31 @@ bool isInLane(
   const std::shared_ptr<lanelet::LaneletMap> & lanelet_map_ptr,
   const geometry_msgs::msg::Point & current_pos)
 {
-  const auto & p = current_pos;
-  const lanelet::Point3d search_point(lanelet::InvalId, p.x, p.y, p.z);
-
-  std::vector<std::pair<double, lanelet::Lanelet>> nearest_lanelets =
-    lanelet::geometry::findNearest(lanelet_map_ptr->laneletLayer, search_point.basicPoint2d(), 1);
```

Comment:
I guess you can use the `nearestUntil` function for this use case, which could be much simpler.
I mean, set a `searchFunc` that filter lanelets by `subtype`.
https://github.com/fzi-forschungszentrum-informatik/Lanelet2/blob/700a96290c9782654e3cf4c287ab1ddec2b7cfbe/lanelet2_examples/src/03_lanelet_map/main.cpp#L138

Could you consider this approach?

Reply:
Sorry for late response. I didn't know that utility. I finally rewrite the function by using `nearestUntil` function. 
https://github.com/autowarefoundation/autoware.universe/pull/824/commits/ced530a857bcfa35f99cb1da1d58293cc27ac2a4

Reply:
If you revert ced530a due to computational cost, I feel it would be acceptable if you refactor (reduce nests) this part.

```cpp
  for (const lanelet::Lanelet & llt : llts_near) {
    if (llt.hasAttribute(lanelet::AttributeName::Subtype)) {
      lanelet::Attribute attr = llt.attribute(lanelet::AttributeName::Subtype);
      if (attr.value() == lanelet::AttributeValueString::Road) {
        if (lanelet::geometry::within(search_point, llt.polygon3d())) {
          return true;
        }
      }
    }
  }
```

I think creating a (lambda) function can make it simpler.
Your current implementation is very clean. :+1: 

```cpp
    // filter only road lanelets
    if (!llt.hasAttribute(lanelet::AttributeName::Subtype)) return false;
    const lanelet::Attribute & attr = llt.attribute(lanelet::AttributeName::Subtype);
    if (attr.value() != lanelet::AttributeValueString::Road) return false;

    // pre-screening out if outside of the bounding box
    const bool is_inside_bbox = lanelet::geometry::within(search_point, llt_box);
    if (!is_inside_bbox) return false;

    // main predicate
    const bool is_inside = lanelet::geometry::within(search_point, llt.polygon2d());
    return is_inside;
```

Reply:
Sorry for late reply. According to https://github.com/fzi-forschungszentrum-informatik/Lanelet2/blob/700a96290c9782654e3cf4c287ab1ddec2b7cfbe/lanelet2_core/src/LaneletMap.cpp#L187 , the nearetUntil function just iterate lanelet element in the nearest order, and check the predicate against them until the nearest element found. So, if I understand correctly, the predicate check will be done for all lanelets in the worst case. 
(writing this comment...)

Reply:
@kenji-miyake 
```
If you revert https://github.com/autowarefoundation/autoware.universe/commit/ced530a857bcfa35f99cb1da1d58293cc27ac2a4 due to computational cost, I feel it would be acceptable if you refactor (reduce nests) this part.
```
Does this mean I don't have to do no additional refactoring after https://github.com/autowarefoundation/autoware.universe/pull/824/commits/ced530a857bcfa35f99cb1da1d58293cc27ac2a4 ?

Reply:
@HiroIshida Thank you for your detailed investigation!

> In the real autonmous driving scene, how many lanelets exist in the map usually?

I guess it's can be 100-1000x larger than the sample map.

> Does this mean I don't have to do no additional refactoring after ced530a ?

With the current implementation, it looks so clean, so you don't have to refactor it anymore.

But if you refactor ced530a by the concern of computational cost, since the nest is deep, I feel you have to refactor code a bit to be as clean as your current predicate function.

Is it clear to you?

Reply:
Yeah, things becomes clear. 
refactor https://github.com/autowarefoundation/autoware.universe/commit/ced530a857bcfa35f99cb1da1d58293cc27ac2a4 is not for computational cost, but for make code clean using search until function.  

```I guess it's can be 100-1000x larger than the sample map.```
If so, current implementation is this PR is not good as it probably takes 0.3 sec or the same order time.

Reply:
Now it's clean code! Thank you!